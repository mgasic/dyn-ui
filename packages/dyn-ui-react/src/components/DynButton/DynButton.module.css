.button {
  /* OSNOVNI STILOVI */
  display: inline-flex;
  align-items: center;
  justify-content: center;
  cursor: pointer;
  border: none;
  text-decoration: none;
  font-family: inherit;
  position: relative;
  transition: var(--animation-transition-all);

  /* TIPOGRAFIJA */
  font-size: var(--size-font-sm);
  font-weight: var(--size-font-weight-medium);
  line-height: var(--size-font-line-height-tight);

  /* SPACING */
  padding: var(--size-spacing-sm) var(--size-spacing-md);
  gap: var(--size-spacing-xs);

  /* BORDER */
  border-radius: var(--size-border-radius-md);
  border-width: var(--size-border-width-sm);
  border-style: solid;

  /* FOKUS STATE */
  &:focus {
    outline: none;
    box-shadow: 0 0 0 var(--size-border-width-md) var(--color-action-focus);
  }

  /* DISABLED STATE */
  &:disabled {
    cursor: not-allowed;
    opacity: var(--animation-opacity-disabled);
  }
}

/* PRIMARY VARIJANTA */
.primary {
  background-color: var(--color-action-default);
  color: var(--color-neutral-light-00);
  border-color: var(--color-action-default);

  &:hover:not(:disabled) {
    background-color: var(--color-action-hover);
    border-color: var(--color-action-hover);
  }

  &:active:not(:disabled) {
    background-color: var(--color-action-pressed);
    border-color: var(--color-action-pressed);
  }
}

/* SECONDARY VARIJANTA */
.secondary {
  background-color: transparent;
  color: var(--color-action-default);
  border-color: var(--color-action-default);

  &:hover:not(:disabled) {
    background-color: var(--color-action-light);
    color: var(--color-action-hover);
    border-color: var(--color-action-hover);
  }

  &:active:not(:disabled) {
    background-color: var(--color-action-pressed-light);
    color: var(--color-action-pressed);
    border-color: var(--color-action-pressed);
  }
}

/* GHOST VARIJANTA */
.ghost {
  background-color: transparent;
  color: var(--color-action-default);
  border-color: transparent;

  &:hover:not(:disabled) {
    background-color: var(--color-action-light);
    color: var(--color-action-hover);
  }

  &:active:not(:disabled) {
    background-color: var(--color-action-pressed-light);
    color: var(--color-action-pressed);
  }
}

/* DANGER VARIJANTA */
.danger {
  background-color: var(--color-feedback-error-default);
  color: var(--color-neutral-light-00);
  border-color: var(--color-feedback-error-default);

  &:hover:not(:disabled) {
    background-color: var(--color-feedback-error-hover);
    border-color: var(--color-feedback-error-hover);
  }

  &:active:not(:disabled) {
    background-color: var(--color-feedback-error-pressed);
    border-color: var(--color-feedback-error-pressed);
  }
}

/* SIZE VARIJANTE */
.small {
  font-size: var(--size-font-xs);
  padding: var(--size-spacing-xs) var(--size-spacing-sm);
  gap: var(--size-spacing-xxs);
}

.medium {
  font-size: var(--size-font-sm);
  padding: var(--size-spacing-sm) var(--size-spacing-md);
  gap: var(--size-spacing-xs);
}

.large {
  font-size: var(--size-font-md);
  padding: var(--size-spacing-md) var(--size-spacing-lg);
  gap: var(--size-spacing-sm);
}

/* FULL WIDTH */
.fullWidth {
  width: 100%;
}

/* LOADING STATE */
.loading {
  .content {
    opacity: 0;
  }

  .spinner {
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
  }
}

/* ICON ONLY */
.iconOnly {
  padding: var(--size-spacing-sm);

  &.small {
    padding: var(--size-spacing-xs);
  }

  &.large {
    padding: var(--size-spacing-md);
  }
}
