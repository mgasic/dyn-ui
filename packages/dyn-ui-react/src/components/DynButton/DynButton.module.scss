/**
 * DynButton SCSS Module - Fixed styling with proper colors and states
 * Following DYN UI design specifications with visual differentiation
 */
@import '../../styles/globals.scss';

// Base button styles
.dynButton {
  @extend .dyn-component;
  
  display: inline-flex;
  align-items: center;
  justify-content: center;
  gap: var(--spacing-sm);
  
  font-family: inherit;
  font-weight: 500;
  text-decoration: none;
  white-space: nowrap;
  cursor: pointer;
  
  border: 1px solid transparent;
  border-radius: var(--radius-md);
  
  transition: var(--transition-colors), var(--transition-transform);
  
  &:focus-visible {
    outline: 2px solid var(--color-primary);
    outline-offset: 2px;
  }
  
  &:hover:not(:disabled) {
    transform: translateY(-1px);
  }
  
  &:active:not(:disabled) {
    transform: translateY(0);
  }
}

// Button kinds (variants) - Fixed with distinct colors
.dynButton--primary {
  background-color: #1976d2; // Blue primary
  color: #ffffff;
  border-color: #1976d2;
  
  &:hover:not(:disabled) {
    background-color: #1565c0; // Darker blue on hover
    border-color: #1565c0;
  }
}

.dynButton--secondary {
  background-color: transparent;
  color: #1976d2; // Blue text
  border-color: #1976d2; // Blue border
  
  &:hover:not(:disabled) {
    background-color: #e3f2fd; // Light blue background on hover
    color: #1565c0;
    border-color: #1565c0;
  }
}

.dynButton--tertiary {
  background-color: transparent;
  color: #1976d2; // Blue text
  border-color: transparent; // No border
  
  &:hover:not(:disabled) {
    background-color: #e3f2fd; // Light blue background on hover
    color: #1565c0;
  }
}

// Danger state - RED colors for Delete, Remove, Cancel
.dynButton--danger {
  &.dynButton--primary {
    background-color: #d32f2f !important; // Red background
    color: #ffffff !important;
    border-color: #d32f2f !important;
    
    &:hover:not(:disabled) {
      background-color: #c62828 !important; // Darker red on hover
      border-color: #c62828 !important;
    }
  }
  
  &.dynButton--secondary {
    color: #d32f2f !important; // Red text
    border-color: #d32f2f !important; // Red border
    
    &:hover:not(:disabled) {
      background-color: #ffebee !important; // Light red background
      color: #c62828 !important;
      border-color: #c62828 !important;
    }
  }
  
  &.dynButton--tertiary {
    color: #d32f2f !important; // Red text
    
    &:hover:not(:disabled) {
      background-color: #ffebee !important; // Light red background
      color: #c62828 !important;
    }
  }
}

// Button sizes
.dynButton--small {
  padding: calc(var(--spacing-xs) * 1.5) var(--spacing-sm);
  font-size: var(--font-size-xs);
  min-height: 28px;
}

.dynButton--medium {
  padding: var(--spacing-sm) var(--spacing-md);
  font-size: var(--font-size-sm);
  min-height: 36px;
}

.dynButton--large {
  padding: var(--spacing-md) var(--spacing-lg);
  font-size: var(--font-size-md);
  min-height: 44px;
}

// States
.dynButton--disabled {
  opacity: 0.6 !important;
  cursor: not-allowed !important;
  pointer-events: none !important;
  
  &:hover {
    transform: none !important;
  }
}

.dynButton--loading {
  pointer-events: none;
  position: relative;
  
  .dynButtonLabel {
    opacity: 0.7;
  }
}

.dynButton--iconOnly {
  min-width: auto;
  aspect-ratio: 1;
}

// Icon styles
.dynButtonIcon {
  display: flex;
  align-items: center;
  justify-content: center;
  flex-shrink: 0;
  
  svg {
    width: 16px;
    height: 16px;
  }
  
  .dynButton--small & svg {
    width: 14px;
    height: 14px;
  }
  
  .dynButton--large & svg {
    width: 18px;
    height: 18px;
  }
}

// Loading spinner
.dynButtonSpinner {
  display: inline-block;
  width: 16px;
  height: 16px;
  border: 2px solid transparent;
  border-top: 2px solid currentColor;
  border-radius: 50%;
  animation: dyn-spin 1s linear infinite;
  flex-shrink: 0;
  margin-right: 8px;
  
  .dynButton--small & {
    width: 14px;
    height: 14px;
  }
  
  .dynButton--large & {
    width: 18px;
    height: 18px;
  }
}

// Label styles
.dynButtonLabel {
  transition: var(--transition-opacity);
  display: flex;
  align-items: center;
  gap: 8px;
}

// Responsive adjustments
@media (max-width: 640px) {
  .dynButton {
    &--small {
      min-height: 32px;
      font-size: var(--font-size-sm);
    }
    
    &--medium {
      min-height: 40px;
      font-size: var(--font-size-md);
    }
    
    &--large {
      min-height: 48px;
      font-size: var(--font-size-lg);
    }
  }
}

// High contrast theme adjustments
.theme-high-contrast .dynButton {
  border-width: 2px;
  
  &:focus-visible {
    outline-width: 3px;
  }
  
  &--tertiary {
    border-color: var(--color-border);
  }
}

// Print styles
@media print {
  .dynButton {
    background: transparent !important;
    color: black !important;
    border: 1px solid black !important;
    transform: none !important;
    
    &--loading .dynButtonSpinner {
      display: none;
    }
  }
}