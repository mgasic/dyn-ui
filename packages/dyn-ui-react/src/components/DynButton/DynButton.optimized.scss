/**
 * DynButton Optimized SCSS - Short semantic class names
 * Replaces long CSS Module hashes with readable short names
 * Final classes: dyn-btn, dyn-btn--primary, dyn-btn--s, etc.
 */

/* Base button - short semantic name */
.dyn-btn {
  display: inline-flex;
  align-items: center;
  justify-content: center;
  gap: 8px;
  border: 1px solid transparent;
  border-radius: 4px;
  font-family: 'Segoe UI', system-ui, sans-serif;
  font-weight: 500;
  text-decoration: none;
  cursor: pointer;
  transition: all 0.2s ease;
  white-space: nowrap;
  position: relative;
  box-sizing: border-box;
  
  &:focus-visible {
    outline: 2px solid #0066cc;
    outline-offset: 2px;
  }
  
  &:hover:not(:disabled) {
    transform: translateY(-1px);
  }
  
  &:active:not(:disabled) {
    transform: translateY(0);
  }
  
  &:disabled {
    opacity: 0.6;
    cursor: not-allowed;
    pointer-events: none;
  }
}

/* Kind variants - single word modifiers */
.dyn-btn--primary {
  background-color: #1976d2;
  color: #ffffff;
  border-color: #1976d2;
  
  &:hover:not(:disabled) {
    background-color: #1565c0;
    border-color: #1565c0;
    box-shadow: 0 2px 8px rgba(25, 118, 210, 0.3);
  }
}

.dyn-btn--secondary {
  background-color: transparent;
  color: #1976d2;
  border-color: #1976d2;
  
  &:hover:not(:disabled) {
    background-color: #e3f2fd;
    color: #1565c0;
    border-color: #1565c0;
  }
}

.dyn-btn--tertiary {
  background-color: transparent;
  color: #1976d2;
  border-color: transparent;
  
  &:hover:not(:disabled) {
    background-color: #e3f2fd;
    color: #1565c0;
  }
}

/* Size variants - single letter modifiers */
.dyn-btn--s { /* small */
  font-size: 12px;
  padding: 6px 12px;
  min-height: 32px;
}

.dyn-btn--m { /* medium */
  font-size: 14px;
  padding: 8px 16px;
  min-height: 40px;
}

.dyn-btn--l { /* large */
  font-size: 16px;
  padding: 12px 20px;
  min-height: 48px;
}

/* State modifiers - short names */
.dyn-btn--danger {
  background-color: #d32f2f;
  color: #ffffff;
  border-color: #d32f2f;
  
  &:hover:not(:disabled) {
    background-color: #c62828;
    border-color: #c62828;
    box-shadow: 0 2px 8px rgba(211, 47, 47, 0.3);
  }
  
  /* Danger combinations */
  &.dyn-btn--secondary {
    background-color: transparent;
    color: #d32f2f;
    border-color: #d32f2f;
    
    &:hover:not(:disabled) {
      background-color: #ffebee;
      color: #c62828;
      border-color: #c62828;
    }
  }
  
  &.dyn-btn--tertiary {
    background-color: transparent;
    color: #d32f2f;
    border-color: transparent;
    
    &:hover:not(:disabled) {
      background-color: #ffebee;
      color: #c62828;
    }
  }
}

.dyn-btn--loading {
  pointer-events: none;
  
  .dyn-btn__content {
    opacity: 0.7;
  }
}

.dyn-btn--icon-only {
  min-width: auto;
  aspect-ratio: 1;
  padding: 8px;
}

.dyn-btn--disabled {
  opacity: 0.6;
  cursor: not-allowed;
  pointer-events: none;
}

/* Element modifiers - BEM style short names */
.dyn-btn__content {
  display: flex;
  align-items: center;
  gap: 8px;
  transition: opacity 0.2s ease;
}

.dyn-btn__icon {
  display: flex;
  align-items: center;
  justify-content: center;
  flex-shrink: 0;
  
  svg {
    width: 16px;
    height: 16px;
    fill: currentColor;
  }
}

.dyn-btn__label {
  display: flex;
  align-items: center;
}

/* Loading spinner - short name */
.dyn-spinner {
  display: inline-block;
  width: 16px;
  height: 16px;
  border: 2px solid transparent;
  border-top: 2px solid currentColor;
  border-radius: 50%;
  animation: dyn-spin 1s linear infinite;
  margin-right: 8px;
  flex-shrink: 0;
}

/* Size specific icon adjustments */
.dyn-btn--s .dyn-btn__icon svg {
  width: 14px;
  height: 14px;
}

.dyn-btn--l .dyn-btn__icon svg {
  width: 18px;
  height: 18px;
}

/* Animations */
@keyframes dyn-spin {
  from { transform: rotate(0deg); }
  to { transform: rotate(360deg); }
}

/* Responsive - maintain short names */
@media (max-width: 640px) {
  .dyn-btn--s {
    min-height: 36px;
    font-size: 14px;
  }
  
  .dyn-btn--m {
    min-height: 44px;
    font-size: 16px;
  }
  
  .dyn-btn--l {
    min-height: 52px;
    font-size: 18px;
  }
}

/* Reduced motion */
@media (prefers-reduced-motion: reduce) {
  .dyn-btn {
    transition: none;
    
    &:hover:not(:disabled) {
      transform: none;
    }
  }
  
  .dyn-spinner {
    animation: none;
  }
}

/* Print styles */
@media print {
  .dyn-btn {
    background: transparent !important;
    color: black !important;
    border: 1px solid black !important;
    transform: none !important;
  }
}