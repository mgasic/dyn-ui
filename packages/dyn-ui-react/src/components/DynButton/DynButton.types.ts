import type {
  ButtonHTMLAttributes,
  FocusEventHandler,
  KeyboardEventHandler,
  MouseEventHandler,
  ReactNode,
} from 'react';
import type { BaseComponentProps } from '../../types';

export type DynButtonKind = 'primary' | 'secondary' | 'tertiary';

export type DynButtonSize = 'small' | 'medium' | 'large';

export interface DynButtonProps
  extends BaseComponentProps,
    Omit<
      ButtonHTMLAttributes<HTMLButtonElement>,
      | 'type'
      | 'onBlur'
      | 'onClick'
      | 'onKeyDown'
      | 'children'
      | 'aria-label'
      | 'aria-expanded'
      | 'aria-controls'
      | 'aria-describedby'
      | 'aria-pressed'
      | keyof BaseComponentProps
    > {
  /** Button text label */
  label?: string;

  /** Icon - can be string (icon name) or React node */
  icon?: string | ReactNode;

  /** HTML button type */
  type?: 'button' | 'submit' | 'reset';

  /** Loading state indicator */
  loading?: boolean;

  /** Text announced to assistive technologies while loading */
  loadingText?: string;

  /** Danger/destructive state styling */
  danger?: boolean;

  /** Button kind/variant */
  kind?: DynButtonKind;

  /** Disabled state */
  disabled?: boolean;

  /** Expand button to full width */
  fullWidth?: boolean;

  /** Hide button on viewports narrower than 768px */
  hideOnMobile?: boolean;

  /** Display as icon-only on mobile while keeping the label accessible */
  iconOnlyOnMobile?: boolean;

  /** Accessible label override */
  ariaLabel?: string;

  /** Accessible expanded state */
  ariaExpanded?: boolean;

  /** ID reference of the element controlled by the button */
  ariaControls?: string;

  /** ID reference of elements describing the button */
  ariaDescribedBy?: string;

  /** Pressed/toggled state for ARIA toggle buttons */
  ariaPressed?: boolean;

  /** Button size */
  size?: DynButtonSize;

  /** Blur event handler */
  onBlur?: FocusEventHandler<HTMLButtonElement>;

  /** Click event handler */
  onClick?: MouseEventHandler<HTMLButtonElement>;

  /** Key down handler */
  onKeyDown?: KeyboardEventHandler<HTMLButtonElement>;
}

export type DynButtonRef = HTMLButtonElement;

export type DynButtonDefaultProps = Required<
  Pick<
    DynButtonProps,
    |
      'type'
      | 'kind'
      | 'size'
      | 'loading'
      | 'loadingText'
      | 'danger'
      | 'disabled'
      | 'fullWidth'
      | 'hideOnMobile'
      | 'iconOnlyOnMobile'
  >
>;

export const DYN_BUTTON_DEFAULT_PROPS: DynButtonDefaultProps = {
  type: 'button',
  kind: 'primary',
  size: 'medium',
  loading: false,
  loadingText: 'Loadingâ€¦',
  danger: false,
  disabled: false,
  fullWidth: false,
  hideOnMobile: false,
  iconOnlyOnMobile: false,
};
