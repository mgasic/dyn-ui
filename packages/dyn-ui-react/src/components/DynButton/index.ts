// Unified exports for DynButton component
export * from './DynButton';
export type { DynButtonProps } from './DynButton.types';