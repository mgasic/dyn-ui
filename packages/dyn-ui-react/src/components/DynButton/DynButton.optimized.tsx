import * as React from 'react';
import { DynButtonProps } from './DynButton.types';
import { cn } from '../../utils/classNames';
import { DynIcon } from '../DynIcon';

/**
 * DynButton - Optimized version with shorter class names
 * Uses semantic class names instead of CSS Module hashes
 */
export const DynButtonOptimized = React.forwardRef<HTMLButtonElement, DynButtonProps>(
  (
    {
      label,
      icon,
      type = 'button',
      loading = false,
      danger = false,
      kind = 'primary',
      disabled = false,
      ariaLabel,
      ariaExpanded,
      size = 'medium',
      className,
      onBlur,
      onClick,
      ...rest
    },
    ref
  ) => {
    const handleClick = () => {
      if (loading || disabled) return;
      onClick?.();
    };

    const handleBlur = () => {
      onBlur?.();
    };

    // Optimized short class names - more readable and smaller bundle
    const buttonClasses = cn(
      // Base class - short and semantic
      'dyn-btn',

      // Kind classes - single word
      `dyn-btn--${kind}`, // dyn-btn--primary, dyn-btn--secondary, dyn-btn--tertiary

      // Size classes - single letter
      `dyn-btn--${size[0]}`, // dyn-btn--s, dyn-btn--m, dyn-btn--l

      // State classes - short abbreviations
      {
        'dyn-btn--danger': danger,
        'dyn-btn--loading': loading,
        'dyn-btn--disabled': disabled,
        'dyn-btn--icon-only': Boolean(icon && !label),
      },

      className
    );

    const renderSpinner = () => {
      if (!loading) return null;
      return (
        <span
          className="dyn-spinner"
          aria-hidden="true"
          data-testid="dyn-button-spinner"
        />
      );
    };

    const renderIcon = () => {
      if (loading) return renderSpinner();
      if (!icon) return null;

      return (
        <DynIcon
          icon={icon}
          size={size}
          className="dyn-btn__icon"
        />
      );
    };

    const renderLabel = () => {
      if (!label) return null;
      return (
        <span className="dyn-btn__label">
          {label}
        </span>
      );
    };

    const effectiveAriaLabel = ariaLabel || (icon && !label ? 'Button' : undefined);

    return (
      <button
        ref={ref}
        type={type}
        className={buttonClasses}
        disabled={disabled || loading}
        aria-label={effectiveAriaLabel}
        aria-expanded={ariaExpanded}
        aria-busy={loading}
        onClick={handleClick}
        onBlur={handleBlur}
        data-testid="dyn-button"
        {...rest}
      >
        {renderIcon()}
        <span className="dyn-btn__content">
          {renderLabel()}
        </span>
      </button>
    );
  }
);

DynButtonOptimized.displayName = 'DynButtonOptimized';

export default DynButtonOptimized;
