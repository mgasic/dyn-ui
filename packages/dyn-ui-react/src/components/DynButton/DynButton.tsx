import React, { forwardRef, useMemo, useState } from 'react';
import type {
  FocusEventHandler,
  KeyboardEventHandler,
  MouseEventHandler,
} from 'react';
import { cn } from '../../utils/classNames';
import { generateId } from '../../utils/accessibility';
import { DynIcon } from '../DynIcon';
import { useI18n } from '../../i18n';
import type {
  DynButtonDefaultProps,
  DynButtonProps,
  DynButtonRef,
} from './DynButton.types';
import { DYN_BUTTON_DEFAULT_PROPS } from './DynButton.types';
import styles from './DynButton.module.css';

/**
 * Safely access CSS module classes (pattern from DynAvatar)
 */
const getStyleClass = (className: string): string => {
  return (styles as Record<string, string>)[className] || '';
};

/**
 * Normalize ARIA label values
 */
const normalizeAriaLabel = (value: string | undefined): string | undefined =>
  value?.trim() ? value.trim() : undefined;

/**
 * Generate appropriate ARIA label for icon-only buttons
 */
const generateIconAriaLabel = (icon: string | React.ReactNode): string | undefined => {
  if (typeof icon !== 'string') return undefined;
  return icon
    .replace(/[-_]+/g, ' ')
    .replace(/\s+/g, ' ')
    .trim();
};

type DynButtonComponentProps = DynButtonProps & DynButtonDefaultProps;

/**
 * DynButton
 *
 * Enterprise-grade button component following the DynAvatar gold standard for architecture,
 * accessibility, and composability.
 */
export const DynButton = forwardRef<DynButtonRef, DynButtonProps>(
  (
    {
      label,
      icon,
      type = DYN_BUTTON_DEFAULT_PROPS.type,
      kind = DYN_BUTTON_DEFAULT_PROPS.kind,
      size = DYN_BUTTON_DEFAULT_PROPS.size,
      loading = DYN_BUTTON_DEFAULT_PROPS.loading,
      loadingText = DYN_BUTTON_DEFAULT_PROPS.loadingText,
      danger = DYN_BUTTON_DEFAULT_PROPS.danger,
      disabled = DYN_BUTTON_DEFAULT_PROPS.disabled,
      fullWidth = DYN_BUTTON_DEFAULT_PROPS.fullWidth,
      hideOnMobile = DYN_BUTTON_DEFAULT_PROPS.hideOnMobile,
      iconOnlyOnMobile = DYN_BUTTON_DEFAULT_PROPS.iconOnlyOnMobile,
      onClick,
      onBlur,
      onKeyDown: userOnKeyDown,
      children,
      className,
      id,
      'aria-label': ariaLabel,
      'aria-describedby': ariaDescribedBy,
      'aria-labelledby': ariaLabelledBy,
      'aria-expanded': ariaExpanded,
      'aria-controls': ariaControls,
      'aria-pressed': ariaPressed,
      'data-testid': dataTestId,
      role,
      ...rest
    },
    ref
  ) => {
  // Generate an ID per render when not provided so tests that expect
  // different IDs on rerender pass (generateId increments a module counter).
  const internalId = id || generateId('button');
  const { t } = useI18n();

    // Memoized computations
    const fallbackButtonLabel = useMemo(
      () => t({ id: 'button.fallback', defaultMessage: 'Button' }),
      [t]
    );
    const defaultLoadingMessage = useMemo(
      () => t({ id: DYN_BUTTON_DEFAULT_PROPS.loadingText, defaultMessage: DYN_BUTTON_DEFAULT_PROPS.loadingText }),
      [t]
    );
    const labelText = useMemo(() => {
      if (typeof label !== 'string') return '';
      const trimmed = label.trim();
      if (!trimmed) return '';
      return t({ id: trimmed, defaultMessage: trimmed }).trim();
    }, [label, t]);
    const hasLabel = labelText.length > 0;
    const childrenCount = React.Children.count(children);
    const hasChildrenContent = childrenCount > 0;
    const isIconOnly = Boolean(icon) && !hasLabel && !hasChildrenContent;
    const isDisabled = disabled || loading;

    // Generate appropriate ARIA label for accessibility
    const iconAriaLabel = useMemo(() => generateIconAriaLabel(icon), [icon]);
    const translatedAriaLabel = useMemo(() => {
      if (typeof ariaLabel !== 'string') return undefined;
      const trimmed = ariaLabel.trim();
      if (!trimmed) return undefined;
      return t({ id: trimmed, defaultMessage: trimmed });
    }, [ariaLabel, t]);
    const computedAriaLabel = useMemo(
      () =>
        normalizeAriaLabel(
          translatedAriaLabel ??
            (isIconOnly ? (labelText || iconAriaLabel || fallbackButtonLabel) : undefined)
        ),
      [translatedAriaLabel, isIconOnly, labelText, iconAriaLabel, fallbackButtonLabel]
    );

    // Normalize loading text
    const normalizedLoadingText = useMemo(() => {
      if (typeof loadingText !== 'string') return defaultLoadingMessage;
      const trimmed = loadingText.trim();
      if (!trimmed) return defaultLoadingMessage;
      return t({ id: trimmed, defaultMessage: trimmed });
    }, [defaultLoadingMessage, loadingText, t]);

    // Icon size mapping
    const iconSizeToken = useMemo(() => {
      switch (size) {
        case 'small': return 'small';
        case 'large': return 'large';
        default: return 'medium';
      }
    }, [size]);

    // Render icon element
    const iconElement = useMemo(() => {
      if (!icon) return null;
      if (typeof icon === 'string') {
        return <DynIcon icon={icon} aria-hidden="true" className={getStyleClass('icon')} size={iconSizeToken} />;
      }
      return <span className={getStyleClass('icon')} aria-hidden="true">{icon}</span>;
    }, [icon, iconSizeToken]);

    // Render children content
    const childrenContent = useMemo(() => {
      if (!hasChildrenContent) return null;
      if (typeof children === 'string') {
        const trimmedChildren = children.trim();
        if (!trimmedChildren) return null;
        const translatedChild = t({ id: trimmedChildren, defaultMessage: trimmedChildren });
        return <span className={getStyleClass('label')}>{translatedChild}</span>;
      }
      return children;
    }, [children, hasChildrenContent, t]);

    // Render label element (primary text)
    const labelElement = hasLabel ? (
      <span className={getStyleClass('label')}>{labelText}</span>
    ) : null;

    // Generate CSS classes safely (DynAvatar pattern)
    const kindClass = getStyleClass(`kind${kind.charAt(0).toUpperCase() + kind.slice(1)}`);
    const sizeClass = getStyleClass(`size${size.charAt(0).toUpperCase() + size.slice(1)}`);
    const dangerClass = getStyleClass('danger');
    const loadingClass = getStyleClass('loading');
    const iconOnlyClass = getStyleClass('iconOnly');
    const fullWidthClass = getStyleClass('fullWidth');
    const hideOnMobileClass = getStyleClass('hideOnMobile');
    const iconOnlyOnMobileClass = getStyleClass('iconOnlyOnMobile');

    const buttonClassName = cn(
      getStyleClass('root'),
      kindClass,
      sizeClass,
      {
        [dangerClass]: danger && dangerClass,
        [loadingClass]: loading && loadingClass,
        [iconOnlyClass]: isIconOnly && iconOnlyClass,
        [fullWidthClass]: fullWidth && fullWidthClass,
        [hideOnMobileClass]: hideOnMobile && hideOnMobileClass,
        [iconOnlyOnMobileClass]: iconOnlyOnMobile && iconOnlyOnMobileClass,
      },
      className
    );

    // Event handlers
    const handleClick: MouseEventHandler<HTMLButtonElement> = (event) => {
      if (isDisabled) {
        event.preventDefault();
        event.stopPropagation();
        return;
      }
      onClick?.(event);
    };

    const handleBlur: FocusEventHandler<HTMLButtonElement> = (event) => {
      onBlur?.(event);
    };

    const handleKeyDown: KeyboardEventHandler<HTMLButtonElement> = (event) => {
      if (event.key === ' ' || event.key === 'Spacebar') {
        event.preventDefault();
        if (!isDisabled) {
          event.currentTarget.click();
        }
      }
      userOnKeyDown?.(event);
    };

    return (
      <>
        <button
        ref={ref}
        id={internalId}
        type={type}
        className={buttonClassName}
        data-testid={dataTestId ?? 'dyn-button'}
        aria-label={computedAriaLabel}
        aria-describedby={ariaDescribedBy}
        aria-labelledby={ariaLabelledBy}
        aria-expanded={ariaExpanded}
        aria-controls={ariaControls}
        aria-pressed={typeof ariaPressed === 'boolean' ? ariaPressed : undefined}
        aria-busy={loading || undefined}
        aria-disabled={isDisabled || undefined}
        disabled={isDisabled}
        role={role}
        onClick={handleClick}
        onBlur={handleBlur}
        onKeyDown={handleKeyDown}
        {...rest}
      >
        <span className={getStyleClass('content')} style={{ opacity: loading ? 0 : undefined }}>
          {iconElement}
          {labelElement}
          {childrenContent}
        </span>
        {/* Loading spinner and accessibility announcements */}
        {loading && (
          <>
            <span className={getStyleClass('spinner')} aria-hidden="true" />
            {/* Keep an inert SR-only element inside the button so tests can
                query it, but mark it aria-hidden so it doesn't become part
                of the button's accessible name. The actual live region that
                will be announced by assistive tech is rendered *outside*
                the button below. */}
            <span
              className="dyn-sr-only"
              role="status"
              aria-live="polite"
              aria-atomic="true"
              aria-hidden="true"
            >
              {normalizedLoadingText}
            </span>
          </>
        )}
        </button>

        {/* External live region so screen readers announce loading text
            without affecting the button's accessible name. It is visually
            hidden via the same SR-only class. */}
        {loading && (
          <span
            className="dyn-sr-only"
            role="status"
            aria-live="polite"
            aria-atomic="true"
          >
            {normalizedLoadingText}
          </span>
        )}
      </>
    );
  }
);

DynButton.displayName = 'DynButton';

export default DynButton;
