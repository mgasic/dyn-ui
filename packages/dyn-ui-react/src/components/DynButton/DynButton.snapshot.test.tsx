import { render } from '@testing-library/react';
import { DynButton } from './DynButton';
import { describe, it, expect } from 'vitest';

describe('DynButton snapshots', () => {
  it('matches visual states snapshot', () => {
    const { container } = render(
      <div style={{ display: 'grid', gap: '0.5rem' }}>
        <DynButton label="Default" />
        <DynButton label="Secondary" variant="secondary" />
        <DynButton label="Danger" danger />
        <DynButton label="Loading" loading />
        <DynButton startIcon="download" aria-label="Download" />
        <DynButton label="End" endIcon="arrow-right" />
      </div>
    );

    expect(container).toMatchSnapshot();
  });
});
