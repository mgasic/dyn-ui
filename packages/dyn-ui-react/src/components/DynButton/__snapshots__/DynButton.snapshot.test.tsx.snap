// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DynButton > matches snapshot 1`] = `
<div>
  <button
    class="_root_c4f507 _kindPrimary_c4f507 _sizeMedium_c4f507"
    data-testid="dyn-button"
    type="button"
  >
    <span
      class="_content_c4f507"
    >
      <span
        class="_label_c4f507"
      >
        Snapshot
      </span>
    </span>
  </button>
</div>
`;
